amy/autoemails/actions.py

```diff
@@ -1235,3 +1245,283 @@ def get_merge_data(self) -> Optional[Dict]:
+    in the last 3 years.
+
+    This action is added during the RQ scheduler startup.
+    See amy.autoemails.management.commands.repeated_jobs
+    """
+
+    launch_at = timedelta(hours=1)
+
+    def recipients(self) -> Optional[Tuple[str]]:
+        """Assuming self.context is ready, overwrite email's recipients
+        with selected ones."""
+        try:
+            person_email = self.context_objects["person_email"]
+            return (person_email,)
+        except (KeyError, AttributeError):
+            return None
+
+    def all_recipients(self) -> str:
+        """If available, return string of all recipients."""
+        recipients = self.recipients()
+        if recipients is not None:
+            return ", ".join(recipients)
+        return ""
+
+    def reply_to(self) -> Optional[Tuple[str]]:
+        """Overwrite in order to set own reply-to from descending Action."""
+        return (settings.ADMIN_NOTIFICATION_CRITERIA_DEFAULT,)
+
+
+class BaseProfileArchivalWarningRepeatedAction(BaseRepeatedAction):
+    """
+    Base class for emails related to profile archival.
+    """
+
+    launch_at: Optional[timedelta] = None
+    REMINDER_LIMIT: int = 3
+    EMAIL_ACTION_CLASS: Optional[Type[BaseAction]] = None
+    ACTION_NAME: Optional[str] = None
+
+    def __call__(self, *args, **kwargs):
+        people = self.get_people()
+        triggers = self.get_triggers()
+        email_reminders = self.get_email_reminders()
+        self.handle_existing_email_reminders(triggers, email_reminders, people)
+        self.create_new_email_reminders(triggers, email_reminders, people)
+
+    def get_people(self) -> Iterable[Person]:
+        """
+        Retrieves the affected users from the database.
+        """
+        return []
+
+    def get_triggers(self) -> Iterable[Trigger]:
+        """
+        Retrieves the related triggers from the database.
+        """
+        return Trigger.objects.filter(
+            active=True,
+            action=self.ACTION_NAME,
+        )
+
+    def get_email_reminders(self) -> Iterable[Trigger]:
+        """
+        Retrieves the related email reminders from the database.
+        """
+        return EmailReminder.objects.filter(
+            archived_at=None,
+            trigger__action=self.ACTION_NAME,
+        ).select_related("person")
+
+    def handle_existing_email_reminders(
+        self,
+        triggers: Iterable[Trigger],
+        email_reminders: Iterable[EmailReminder],
+        people: Iterable[Person],
+    ) -> None:
+        """
+        Update existing email reminders doing the following:
+        - Archive the email reminder where the users have logged in.
+        - Send another reminder if the user has not yet hit the REMINDER_LIMIT
+        - If the user hasn't logged in and they have hit the REMINDER_LIMIT,
+          archive the user and the email reminder.
+        """
+        people_ids = set([person.id for person in people])
```

Comment:
@pbanaszkiewicz 
I don't believe a generator is applicable here. Later in the loop I am doing:
`if reminder.person_id not in people_ids:`

The set `people_ids` must evaluate in full on that line so that you can determine if the given person_id is in the set. Alternatively if you're concerned about storing all potentially archivable people_ids in memory (I would assume there weren't that many but I could be mistaken I haven't looked at the data), then it's best that this is done as a query on the database side. 

Let me know if you think it's unreasonable to hold them in memory or if I misinterpreted your comment.

Reply:
@lauryndbrown I meant "use generator expression instead of list comprehension for creating the set". Sorry for not stating it clearly.

You can switch `set([person.id for person in people])` to `set(person.id for person in people)` to eliminate the additional list.

Reply:
Oh thanks for letting me know.